include/xtensor/xmasked_value.hpp

```diff
@@ -0,0 +1,817 @@
+        using flag_type = B;
+        using optional_type = xtl::xoptional<value_type, flag_type>;
+
+        template <class T1, class B1>
+        inline constexpr xmasked_value(T1&& value, B1&& flag)
+            : m_value(std::forward<T1>(value)), m_flag(std::forward<B1>(flag))
+        {
+        }
+
+        inline explicit constexpr xmasked_value(const xtl::xoptional<T, B>& opt)
+            : m_value(opt.value()), m_flag(opt.has_value())
+        {
+        }
+
+        inline operator value_type() {
+            return m_value;
+        }
+
+        inline operator optional_type() {
+            return optional_type(m_value, m_flag);
+        }
+
+        inline value_type value()
+        {
+            return m_value;
+        }
+
+        inline flag_type has_value()
+        {
+            return m_flag;
+        }
+
+        inline value_type value() const
+        {
+            return m_value;
+        }
+
+        inline flag_type has_value() const
+        {
+            return m_flag;
+        }
+
+        template <class T1, class B1>
+        inline bool equal(const xmasked_value<T1, B1>& rhs) const noexcept
+        {
+            return (!m_flag && !rhs.m_flag) || (m_value == rhs.m_value && (m_flag && rhs.m_flag));
+        }
+
+        template <class T1, class B1>
+        inline bool equal(const xtl::xoptional<T1, B1>& rhs) const noexcept
+        {
+            return (!m_flag && !rhs.has_value()) || (m_value == rhs.value() && (m_flag && rhs.has_value()));
+        }
+
+        template <class T1>
+        inline auto equal(const T1& rhs) const noexcept -> disable_xoptional_like<T1, bool>
+        {
+            return m_flag && m_value == rhs;
+        }
+
+#define DEFINE_ASSIGN_OPERATOR(OP)                                                            \
+        template <class T1>                                                                   \
+        inline auto operator OP(const T1& value) -> disable_xoptional_like<T1, xmasked_value&>\
+        {                                                                                     \
+            if (m_flag)                                                                       \
+            {                                                                                 \
+                m_value OP value;                                                             \
+            }                                                                                 \
+            return *this;                                                                     \
+        }                                                                                     \
+                                                                                              \
+        template <class T1, class B1>                                                         \
+        inline xmasked_value& operator OP(const xtl::xoptional<T1, B1>& value)                \
```

Comment:
Remark global to operator overloads: since the mask of the `xmasked_value` is propagated to the `xoptional` in the conversion operator, no need to handle special overloads taking `xoptional` arguments.